File: server/src/main/java/backend/config/SwaggerConfig.java
package backend.config;

import backend.controller.swagger.request.PageableRequest;
import com.fasterxml.classmate.TypeResolver;
import lombok.RequiredArgsConstructor;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.data.domain.Pageable;
import springfox.documentation.builders.ApiInfoBuilder;
import springfox.documentation.builders.RequestHandlerSelectors;
import springfox.documentation.schema.AlternateTypeRules;
import springfox.documentation.service.*;
import springfox.documentation.spi.DocumentationType;
import springfox.documentation.spi.service.contexts.SecurityContext;
import springfox.documentation.spring.web.plugins.Docket;
import springfox.documentation.swagger2.annotations.EnableSwagger2;

import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;


@Configuration
@RequiredArgsConstructor
@EnableSwagger2
public class SwaggerConfig {
    private final TypeResolver typeResolver;
    private final String API_VERSION = "2.1";

    @Bean
    public Docket sabotenApi() {
        return new Docket(DocumentationType.SWAGGER_2)
                .alternateTypeRules(AlternateTypeRules
                        .newRule(typeResolver.resolve(Pageable.class), typeResolver.resolve(PageableRequest.class)))
                .consumes(getConsumeContentTypes())
                .produces(getProduceContentTypes())
                .select()
                .apis(RequestHandlerSelectors.any())
                //.apis(RequestHandlerSelectors.basePackage("server.src.main.java.backend.controller"))
                .build()
                .apiInfo(this.metaInfo())
                .securityContexts(Arrays.asList(securityContext()))
                .securitySchemes(Arrays.asList(apiKey()));
    }

    private ApiInfo metaInfo() {
        return new ApiInfoBuilder()
                .title("Project Saboten")
                .description("GDSC-DJU 선인장 (선택장애 인간들을 위한 장) 프로젝트")
                .version(API_VERSION)
                .termsOfServiceUrl("https://github.com/GDSC-Daejin/project-saboten.git")
                .license("MIT")
                .licenseUrl("https://license.com")
                .contact(
                        new Contact(
                                "Contact Us",
                                "https://github.com/GDSC-Daejin/project-saboten.git",
                                "shinequasar@naver.com"
                        )
                )
                .build();
    }

    private Set<String> getConsumeContentTypes() {
        Set<String> consumes = new HashSet<>();
        consumes.add("application/json;charset=UTF-8");
        consumes.add("application/x-www-form-urlencoded");
        return consumes;
    }

    private Set<String> getProduceContentTypes() {
        Set<String> produces = new HashSet<>();
        produces.add("application/json;charset=UTF-8");
        return produces;
    }

    private SecurityContext securityContext() {
        return SecurityContext.builder().securityReferences(defaultAuth()).build();
    }

    private List<SecurityReference> defaultAuth() {
        AuthorizationScope authorizationScope = new AuthorizationScope("global", "accessEverything");
        AuthorizationScope[] authorizationScopes = new AuthorizationScope[1];
        authorizationScopes[0] = authorizationScope;
        return Arrays.asList(new SecurityReference("JWT", authorizationScopes));
    }

    private ApiKey apiKey() {
        return new ApiKey("JWT", "Authorization", "header");
    }
}

File: server/src/main/java/backend/docs/SwaggerConfig.kt
package backend.docs

import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import springfox.documentation.builders.ApiInfoBuilder
import springfox.documentation.builders.RequestHandlerSelectors
import springfox.documentation.service.ApiInfo
import springfox.documentation.service.Contact
import springfox.documentation.spi.DocumentationType
import springfox.documentation.spring.web.plugins.Docket
import springfox.documentation.swagger2.annotations.EnableSwagger2

private const val API_VERSION = "1.0"

@Configuration
@EnableSwagger2
class SwaggerConfig {

    @Bean
    fun sabotenApi(): Docket {
        return Docket(DocumentationType.SWAGGER_2)
            .select()
            .apis(RequestHandlerSelectors.any())
            //.apis(RequestHandlerSelectors.basePackage("server.src.main.java.backend.controller"))
            .build()
            .apiInfo(this.metaInfo())
    }

    private fun metaInfo(): ApiInfo {
        return ApiInfoBuilder()
            .title("Project Saboten")
            .description("GDSC-DJU 선인장 (선택장애 인간들을 위한 장) 프로젝트")
            .version(API_VERSION)
            .termsOfServiceUrl("https://github.com/GDSC-Daejin/project-saboten.git")
            .license("MIT")
            .licenseUrl("https://license.com")
            .contact(
                Contact(
                    "Contact Us",
                    "https://github.com/GDSC-Daejin/project-saboten.git",
                    "shinequasar@naver.com"
                )
            )
            .build()
    }




}